checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInferenceLogic.java

```diff
@@ -0,0 +1,164 @@
+import org.checkerframework.checker.mustcall.qual.Owning;
+import org.checkerframework.common.wholeprograminference.WholeProgramInference;
+import org.checkerframework.dataflow.cfg.ControlFlowGraph;
+import org.checkerframework.dataflow.cfg.block.Block;
+import org.checkerframework.dataflow.cfg.block.ConditionalBlock;
+import org.checkerframework.dataflow.cfg.block.SingleSuccessorBlock;
+import org.checkerframework.dataflow.cfg.node.MethodInvocationNode;
+import org.checkerframework.dataflow.cfg.node.Node;
+import org.checkerframework.javacutil.AnnotationBuilder;
+import org.checkerframework.javacutil.BugInCF;
+import org.checkerframework.javacutil.TreeUtils;
+
+/**
+ * This class contains Resource Leak Checker annotation inference algorithm. For now, it just
+ * contains inference logic for owning annotations on final owning fields. It adds an @Owning
+ * annotation on a field if it finds a method that satisfies @MustCall obligation of the field along
+ * some path to the regular exit point.
+ */
+public class MustCallInferenceLogic {
+
+  /** The set of owning fields. */
+  private Set<Element> owningFields = new HashSet<>();
+  /**
+   * The type factory for the Resource Leak Checker, which is used to access the Must Call Checker.
+   */
+  private final ResourceLeakAnnotatedTypeFactory typeFactory;
+
+  /** The @Owning annotation. */
+  protected final AnnotationMirror OWNING;
+
+  /** The control flow graph. */
+  private ControlFlowGraph cfg;
+
+  /**
+   * Creates a MustCallInferenceLogic. If the whole program inference is not null, the type
+   * factory's postAnalyze method would instantiate a new MustCallInferenceLogic using this
+   * constructor and then call {@link #runInference()}.
+   *
+   * @param typeFactory the type factory
+   * @param cfg the ControlFlowGraph
+   */
+  MustCallInferenceLogic(ResourceLeakAnnotatedTypeFactory typeFactory, ControlFlowGraph cfg) {
+    this.typeFactory = typeFactory;
+    this.cfg = cfg;
+    OWNING = AnnotationBuilder.fromClass(this.typeFactory.getElementUtils(), Owning.class);
+  }
+
+  /**
+   * It checks method invocations for fields with non-empty @MustCall obligation along all paths to
+   * the regular exit point in the method body.
+   */
+  void runInference() {
+    Set<Block> visited = new HashSet<>();
+    Deque<Block> worklist = new ArrayDeque<>();
+    Block entry = this.cfg.getEntryBlock();
+
+    worklist.add(entry);
+    visited.add(entry);
+
+    while (!worklist.isEmpty()) {
+      Block current = worklist.remove();
+
+      for (Node node : current.getNodes()) {
+        if (node instanceof MethodInvocationNode) {
+          checkForMustCallInvocationOnField((MethodInvocationNode) node);
+        }
+      }
+
+      propagateRegPaths(current, visited, worklist);
+    }
+  }
+
+  /**
+   * If the receiver of {@code mNode} is a possible owning field and the method invocation satisfies
+   * the field's must call obligation, then adds owning annotation for that field.
```

Comment:
This documentation is out of date.  The method actually marks the field as owning; the annotation is added later